File: frontend/src/services/research.ts
/**
 * 研究 API 封装
 */

import { apiFetch } from './http';
import { openSseStream } from './sse';
import type { SseEvent } from '../lib/sse';
import type {
  ResearchArtifactsResponse,
  ResearchClarificationSubmitRequest,
  ResearchPlanUpdateRequest,
  ResearchSessionAccepted,
  ResearchSessionCreateRequest,
  ResearchStopRequest,
} from '../types/researchEvents';

// Deep Research 的预规划接口会等待 scoper 判定，并停在澄清/计划确认阶段；
// 30 秒默认超时过短，用户在提交补充信息或更新计划后容易被前端提前中断。
const RESEARCH_PLANNING_TIMEOUT_MS = 300_000;

export type {
  ResearchArtifactRead,
  ResearchArtifactsResponse,
  ResearchCanonicalCitation,
  ResearchClarificationQuestion,
  ResearchClarificationRequest,
  ResearchClarificationSubmitRequest,
  ResearchEventEnvelope,
  ResearchPlanUpdateRequest,
  ResearchPlanSnapshot,
  ResearchPlanSubtask,
  ResearchSessionAccepted,
  ResearchSessionCreateRequest,
  ResearchSessionStatus,
  ResearchSessionView,
  ResearchStopRequest,
  ResearchSourceTarget,
  ResearchSourceType,
} from '../types/researchEvents';

export interface ResearchStreamOptions {
  signal?: AbortSignal;
  lastEventId?: string | null;
  resumeFromEventId?: string | null;
}

export function buildResearchStreamPath(
  sessionId: string,
  resumeFromEventId?: string | null
): string {
  const path = `/api/v1/research/sessions/${sessionId}/stream`;
  if (!resumeFromEventId) {
    return path;
  }

  const params = new URLSearchParams({
    resume_from_event_id: resumeFromEventId,
  });
  return `${path}?${params.toString()}`;
}

/**
 * 发起深度研究会话
 */
export async function createResearchSession(
  data: ResearchSessionCreateRequest
): Promise<ResearchSessionAccepted> {
  const payload: ResearchSessionCreateRequest = {
    question: data.question.trim(),
    plan_first: true,
  };

  return apiFetch<ResearchSessionAccepted>('/api/v1/research/sessions', {
    method: 'POST',
    body: JSON.stringify(payload),
    timeoutMs: RESEARCH_PLANNING_TIMEOUT_MS,
  });
}

/**
 * 提交澄清回答
 */
export async function submitResearchClarification(
  sessionId: string,
  data: ResearchClarificationSubmitRequest
): Promise<ResearchSessionAccepted> {
  return apiFetch<ResearchSessionAccepted>(
    `/api/v1/research/sessions/${sessionId}/clarification`,
    {
      method: 'POST',
      body: JSON.stringify(data),
      timeoutMs: RESEARCH_PLANNING_TIMEOUT_MS,
    }
  );
}

export async function updateResearchPlan(
  sessionId: string,
  data: ResearchPlanUpdateRequest
): Promise<ResearchSessionAccepted> {
  return apiFetch<ResearchSessionAccepted>(`/api/v1/research/sessions/${sessionId}/plan`, {
    method: 'POST',
    body: JSON.stringify(data),
    timeoutMs: RESEARCH_PLANNING_TIMEOUT_MS,
  });
}

export async function startResearchSession(sessionId: string): Promise<ResearchSessionAccepted> {
  return apiFetch<ResearchSessionAccepted>(`/api/v1/research/sessions/${sessionId}/start`, {
    method: 'POST',
  });
}

/**
 * 获取研究工件
 */
export async function getResearchArtifacts(
  sessionId: string
): Promise<ResearchArtifactsResponse> {
  return apiFetch<ResearchArtifactsResponse>(
    `/api/v1/research/sessions/${sessionId}/artifacts`
  );
}

/**
 * 中断研究会话
 */
export async function stopResearchSession(
  sessionId: string,
  data: ResearchStopRequest = {}
): Promise<ResearchSessionAccepted> {
  return apiFetch<ResearchSessionAccepted>(`/api/v1/research/sessions/${sessionId}/stop`, {
    method: 'POST',
    body: JSON.stringify(data),
  });
}

/**
 * 研究事件流
 */
export async function streamResearchSession(
  sessionId: string,
  options: ResearchStreamOptions = {}
): Promise<AsyncIterable<SseEvent>> {
  const headers: Record<string, string> = {};
  if (options.lastEventId) {
    headers['Last-Event-ID'] = options.lastEventId;
  }

  return openSseStream(
    buildResearchStreamPath(sessionId, options.resumeFromEventId),
    {
      method: 'GET',
      ...(Object.keys(headers).length > 0 ? { headers } : {}),
    },
    options.signal
  );
}

File: backend/src/app/api/v1/endpoints/research.py
"""当前 research 会话端点。"""

from __future__ import annotations

import uuid
from collections.abc import AsyncIterator, Sequence

from fastapi import APIRouter, Request, status
from fastapi.responses import StreamingResponse

from app.api.dependencies.services import ResearchServiceDep
from app.api.deps import AsyncSessionDep
from app.api.sse import SSE_HEADERS, encode_sse
from app.schemas.research import (
    ResearchArtifactsResponse,
    ResearchClarificationSubmitRequest,
    ResearchEventEnvelope,
    ResearchPlanUpdateRequest,
    ResearchSessionAccepted,
    ResearchSessionCreateRequest,
    ResearchStopRequest,
    ResearchStreamResumeParams,
)

router = APIRouter()


async def _emit_research_events(
    *,
    envelopes: Sequence[ResearchEventEnvelope],
) -> AsyncIterator[tuple[str, object]]:
    for envelope in envelopes:
        yield "research.event", envelope.model_dump(mode="json")


@router.post(
    "/sessions",
    response_model=ResearchSessionAccepted,
    status_code=status.HTTP_201_CREATED,
)
async def create_research_session(
    db: AsyncSessionDep,
    service: ResearchServiceDep,
    body: ResearchSessionCreateRequest,
) -> ResearchSessionAccepted:
    session_id = uuid.uuid4()
    session, plan_result = await service.create_session(
        body,
        session_id=session_id,
        thread_id=str(session_id),
    )
    await db.commit()
    return ResearchSessionAccepted(
        session_id=session.id,
        question=session.question,
        status=session.status,
        plan_snapshot=plan_result.plan_snapshot,
        clarification_request=plan_result.clarification_request,
    )


@router.post(
    "/sessions/{session_id}/clarification",
    response_model=ResearchSessionAccepted,
)
async def submit_research_clarification(
    session_id: uuid.UUID,
    db: AsyncSessionDep,
    service: ResearchServiceDep,
    body: ResearchClarificationSubmitRequest,
) -> ResearchSessionAccepted:
    session = await service.get_session(session_id)
    session, plan_result = await service.submit_clarification(
        session=session,
        answer=body.answer,
    )
    await db.commit()
    return ResearchSessionAccepted(
        session_id=session.id,
        question=session.question,
        status=session.status,
        plan_snapshot=plan_result.plan_snapshot,
        clarification_request=plan_result.clarification_request,
    )


@router.post(
    "/sessions/{session_id}/plan",
    response_model=ResearchSessionAccepted,
)
async def update_research_plan(
    session_id: uuid.UUID,
    db: AsyncSessionDep,
    service: ResearchServiceDep,
    body: ResearchPlanUpdateRequest,
) -> ResearchSessionAccepted:
    session = await service.get_session(session_id)
    session, plan_result = await service.update_plan(
        session=session, feedback=body.feedback
    )
    await db.commit()
    return ResearchSessionAccepted(
        session_id=session.id,
        question=session.question,
        status=session.status,
        plan_snapshot=plan_result.plan_snapshot,
        clarification_request=plan_result.clarification_request,
    )


@router.post(
    "/sessions/{session_id}/start",
    response_model=ResearchSessionAccepted,
)
async def start_research_session(
    session_id: uuid.UUID,
    db: AsyncSessionDep,
    service: ResearchServiceDep,
) -> ResearchSessionAccepted:
    session = await service.get_session(session_id)
    session = await service.start_session(session=session)
    await db.commit()
    service.trigger_outbox_dispatch()
    return ResearchSessionAccepted(
        session_id=session.id,
        question=session.question,
        status=session.status,
        plan_snapshot=service.read_plan_snapshot(session),
        clarification_request=None,
    )


@router.get("/sessions/{session_id}/stream")
async def stream_research_session(
    session_id: uuid.UUID,
    service: ResearchServiceDep,
    request: Request,
    resume_from_event_id: str | None = None,
) -> StreamingResponse:
    session = await service.get_session(session_id)
    resume_params = ResearchStreamResumeParams(
        resume_from_event_id=resume_from_event_id
    )
    after_event_id = resume_params.effective_after_event_id(
        last_event_id=request.headers.get("Last-Event-ID")
    )
    envelopes = service.list_event_envelopes(
        session,
        after_event_id=after_event_id,
    )
    stream = encode_sse(
        _emit_research_events(
            envelopes=envelopes,
        )
    )
    return StreamingResponse(
        stream,
        media_type="text/event-stream",
        headers=SSE_HEADERS,
    )


@router.post(
    "/sessions/{session_id}/stop",
    response_model=ResearchSessionAccepted,
)
async def stop_research_session(
    session_id: uuid.UUID,
    db: AsyncSessionDep,
    service: ResearchServiceDep,
    body: ResearchStopRequest,
) -> ResearchSessionAccepted:
    session = await service.get_session(session_id)
    session = await service.stop_session(session=session, reason=body.reason)
    await db.commit()
    return ResearchSessionAccepted(
        session_id=session.id,
        question=session.question,
        status=session.status,
    )


@router.get(
    "/sessions/{session_id}/artifacts",
    response_model=ResearchArtifactsResponse,
)
async def list_research_artifacts(
    session_id: uuid.UUID,
    service: ResearchServiceDep,
) -> ResearchArtifactsResponse:
    session = await service.get_session(session_id)
    return service.build_artifacts_response(session)
